RFC: reverse sstable reader

wmitros opened the pull request:
You can find detailed description in the commit messages.

Based on https://github.com/denesb/scylla/tree/reverse-reads/v2

wmitros commented:
The changes passed dev unit tests, but they definitely require further testing

avikivity commented on clustering_bounds_comparator.hh:
```diff
@@ -136,6 +136,20 @@ public:
     }
     template<template<typename> typename R>
+    requires Range<R, clustering_key_prefix_view>
+    static bound_view from_range_reverse_start(const R<clustering_key_prefix>& range) {
```
What values can R take apart from `interval`?

avikivity replied:
Ah, I see you just extended the existing practice.

avikivity commented on clustering_ranges_walker.hh:
```diff
@@ -69,13 +74,18 @@ private:
+                } else {
+                    _current_start = position_in_partition_view::for_range_start(_current_range.front());
+                    _current_end = position_in_partition_view::for_range_end(_current_range.front());
+                }
```
I think we'll need some words to explain things, because this is confusing. IIUC _current_range is reversed, and you're un-reversing it.

Let's call the clustering keys as passed in from the query as in the "query domain", and the unreversed keys and comparators the "sstable domain". Without reversed queries, they are the same. With reversed queries, they aren't. We can start by adding comments about what domain a function input or output or variable is in, and later perhaps change the variable and function names too. I wouldn't go as far as to use the type system, but maybe it's merited.

wmitros replied:
The confusion around `position_in_partition_view` mostly arises from the fact, that when using the reversed schema, the comparisons on clustering key prefixes are reversed, but the comparisons on `bound_weight` remain the same. 
We treat a key as included in a range if it's not less than the beginning of the range, but it is less than the end of the range. As a result, if we had a clustering range including keys from ck1 to ck2, we would have a beginning position_in_partition_view looking like (ck1, before), and an ending like (ck2, after). If we want to reverse this range so that the comparisons work with the reversed schema, we can't simply swap the beginning and the end (because the comparisons on "before" and "after" weren't reversed), but use a start (ck2, before) and (ck1, after) instead.

In this case `_current_range` is actually in the "query domain" and we're simply setting correct `bound_weight`s for the reversed range, using the end key as a start, and the start key as end.

avikivity replied:
(and yes, potentially reversed == query domain, unreversed == sstable/table domain)

wmitros replied:
Ok

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+
+namespace mx {
+
+class partition_header_context : public data_consumer::continuous_data_consumer<partition_header_context> {
```
Please add some words what this thing is.

avikivity replied:
Beyond the general description, the mapping from the original sstable data file offsets to the reversed data stream should be explained (if it's really needed, I assume so because the index outputs offsets).

avikivity replied:
Ah, the changelog says we don't do that. I'll read it carefully.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+    }
+    processing_result_generator do_process_state() {
+        co_yield read_16(*_processing_data);
+        co_yield skip(*_processing_data, uint32_t{_u16} + sizeof(uint32_t) + sizeof(uint64_t));
```
Do leave comments naming the fields you skip, to give anyone who reads this code some context.

wmitros replied:
Ok

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+                _reading_range_tombstone_ck = false;
+            }
+            co_yield read_unsigned_vint(*_processing_data);
+            _next_row_offset = current_position() + _u64;
```
Idea (for later): extend yield support to allow writing ``next_row_offset = current_position + co_yield read_unsigned_vint();`.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+
+    virtual future<temporary_buffer<char>> get() override {
+        if (!_partition_header_context) {
+            _partition_header_context.emplace(data_stream(_partition_start, _partition_end), _partition_start, _partition_end - _partition_start, _permit);
```
Perhaps use information from the promoted index to reduce the stream size, can be used to reduce overreads. Minor, we can do it in a later pass.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+            // our way of noticing a new clustering range
+            _row_start = *_ir.data_file_positions().end;
+            _state = state::RANGE_END;
+        }
```
How does fast_forward_to(position_range) work? It interacts with the index directly, so it works with non-reversed offsets.

wmitros replied:
Did it clarify [here](https://github.com/scylladb/scylla/pull/9171#discussion_r685806342)?
According to https://github.com/scylladb/scylla/blob/4ae6eae00a90ecf7cccf1ec7a2bd45cd06a2344b/flat_mutation_reader_v2.hh#L394-L398
fast_forward_to(position_range) is not supposed to be used on single partition readers, so it does not work with reverse reads right either
We could probably modify it so that it works, but I'm not even sure if the range that we're forwarding to is reversed and I can't really find places where we're using this, even on multiple partition reads (clustering_order_reader_merger::peek_reader looks like a candidate)

tgrabiec replied:
I think that comment only applies to the `partition_range` variant. 

Btw, we should eventually respect the fwd_mr flag in deciding whether f_f_t() across partitions is supported. The problem is that it's yes by default, so we would lose the single-partition optimization, which benefits from not having to worry about wider partition range. There were patches on the mailing list to fix that but they caused some unit test failures which were not investigated. 

The range passed to reader's f_f_t() should conform to the reader's schema, so reversed. Reversing reader wrapper should intercept this and flip.

haaawk commented on clustering_key_filter.hh:
```diff
@@ -63,6 +63,10 @@ public:
         }
         return clustering_key_filter_ranges(ranges);
     }
+    static clustering_key_filter_ranges get_maybe_reversed_ranges(const schema& schema, const query::partition_slice& slice, const partition_key& key) {
```
Why is this method called `maybe`? It never reverses, does it?

wmitros replied:
It does not reverse the order of ranges, but in case we have `reversed` in the `partition_slice`, the order of ranges has been actually reversed beforehand, so if we do not reverse them again, we actually return them in reversed order

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,442 @@
+            return _cached_read.size() - (_row_end - offset);
+        };
+        _cached_read.get_write()[to_idx(info.kind_offset)] = 7 - (int)info.range_tombstone_kind;
+        if ((int)info.range_tombstone_kind < 2 || (int)info.range_tombstone_kind > 5) {
```
A lot more comments are needed here.

wmitros replied:
I think I'll handle this with a switch instead of converting the enums to ints

avikivity commented on sstables/mx/reader.cc:
```diff
@@ -1373,22 +1373,35 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
                     _sst->get_stats().on_partition_seek();
-                    auto open_end_marker = idx.end_open_marker();
+                    auto open_end_marker = _index_reader->end_open_marker();
                     if (open_end_marker) {
```
Okay, so this is how fast forwarding is done.

avikivity commented:
Looks good. We can run the regular sstable tests with the double-reversing reader.

haaawk commented on clustering_ranges_walker.hh:
```diff
@@ -313,6 +323,9 @@ public:
 
     // Returns the upper bound of the last range in provided ranges set
     position_in_partition_view uppermost_bound() const {
```
It seems that this function is only called from mutation_fragment_filter::uppermost_bound which is never used so maybe it's easier to just remove it. I will send a patch.

wmitros commented on sstables/mx/reader.cc:
```diff
@@ -1373,22 +1373,35 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
                     if (open_end_marker) {
                         _consumer.set_range_tombstone(open_end_marker->tomb);
                     } else {
                         _consumer.set_range_tombstone({});
                     }
```
I was kind of leaving the handling of the end_open_marker for later, but I just noticed something.
Is it possible that we don't need to change anything concerning the end_open_marker?
After the changes we need something like a start_open_marker that we would get from the promoted_index_block following our current range end (or start if we're talking about reversed ranges). But if I understand correctly, the start_open_marker from the next block would be the same tombstone, as the end_open_marker from the previous one, so because we advance the index to the range end, get the tombstone that we want.

avikivity commented on clustering_ranges_walker.hh:
```diff
@@ -58,8 +58,13 @@ private:
         }
-        _current_start = position_in_partition_view::for_range_start(_current_range.front());
-        _current_end = position_in_partition_view::for_range_end(_current_range.front());
+        if (_schema.current_transformation() == schema_transformation_type::reversed) {
```
Yes. I think that needs to be done in storage_proxy::query_mutations_locally() and storage_proxy::query_result_local(), as entry points into the replica.

wmitros replied:
Now that I look at this, if we reverse the ranges on a higher level, we will need to un-reverse them for the kl sstable reader. Also, do we want to have them reversed at the memtable/sstable reader level (we're passing unreversed schemas to them)?

avikivity replied:
I think we do. The reason is to keep everything in the same domain (the query domain). If different bits of the parameters use different domains, it's much harder to keep track.

The k/l reader works only at the table domain and we use the reversing reader to reverse it. So it makes sense for the reversing reader to transform the inputs from the query domain to the table domain, then the k/l reader has identical query and table domains.

wmitros commented:
I've added quite some comments to the changes, but didn't update the reversal of the ranges in a higher layer than the clustering_ranges_walker yet.

wmitros commented:
Before passing this patch to @haaawk, as I'm going on vacation, I've fixed the issues that I was able to find until now, the last rebase includes 3 important changes:
* when skipping between ranges, we're skipping to the closest promoted index block from them all, instead of only using blocks that we had cached - we often didn't have any blocks that we wanted, so we were using the end of partition as the range start, which was particularly slow because we then needed to find the last row of the last block. Besides, it's more efficient to perform one read from index to skip at least one additional block, than to read the entire block from datafile. There is a non-obvious detail about the new skipping method, as it accepts reversed positions, but uses non-reversed schema on its own (advance_past)
* the intermediary cache was not adjusted after skipping between ranges - it's fixed now
* in the intermediary stream, we're producing row by row, so we want the consumer to read entire rows. Before this patch we were sometimes reading row start and then skipping based on its clustering key. Now, when in a reverse read, we're always finishing reading the row, possibly not emitting the corresponding mutation_fragment
Minor fixes include:
* some comments
* fixed naming of some variables (used "entry", while the correct term was "block")
* used the end open marker from the upper bound of the index, instead of the lower
* handled the empty partition case (only partition tombstone) and the no-block case (currently, if we only had 1 promoted index block, we have 0 instead)

wmitros commented:
This was adapted in https://github.com/scylladb/scylladb/pull/9281/, so we can close this draft as well
